src/ERC721/ERC721/ERC721Facet.sol

```diff
@@ -239,4 +239,86 @@ contract ERC721Facet {
             }
         }
     }   
+
+    /// @notice Provide the metadata URI for a given token ID.
+    /// @param _tokenId tokenID of the NFT to query the metadata from
+    /// @return the URI providing the detailled metadata of the specified tokenID
+    function tokenURI(uint256 _tokenId) external view returns (string memory) {
+        ERC721Storage storage s = getStorage();
+        address owner = s.ownerOf[_tokenId];
+        if (owner == address(0)) {
+            revert ERC721NonexistentToken(_tokenId);
+        }
+
+        return bytes(s.baseURI).length > 0 ? string.concat(s.baseURI, toString(_tokenId)) : "";
+    }
+
+    /// @notice Provide the metadata URI for a given token ID.
+    /// @param _newBaseURI The common base URI path for all token metadatas.
+    function setBaseURI(string calldata _newBaseURI) external {
+        ERC721Storage storage s = getStorage();
+        s.baseURI = _newBaseURI;
+    }
+
+    /// From openzeppelin/contracts/utils/Strings.sol
+    /// @dev Converts a `uint256` to its ASCII `string` decimal representation.
+    function toString(uint256 value) internal pure returns (string memory) {
+        bytes16 _SYMBOLS = "0123456789abcdef";
+        unchecked {
+            uint256 length = log10(value) + 1;
+            string memory buffer = new string(length);
+            uint256 ptr;
+            /// @solidity memory-safe-assembly
+            assembly {
```

Comment:
Instead of using `/// @solidity memory-safe-assembly` before the assembly statement, remove the comment and use this instead: 

```solidity
assembly ("memory-safe") {
```

This is covered in Solidity documentation here: https://docs.soliditylang.org/en/v0.8.30/assembly.html#memory-safety

Using the comment is the old way.